src/toil/cwl/cwltoil.py

```diff
@@ -937,35 +1539,64 @@ def _collectDirEntries(
-        if p.staged:
-            if destBucket and p.type in ["File", "CreateFile"]:
-                # Directories don't need to be created if we're exporting to a bucket
-                baseName = p.target[len(outdir) :]
-                local_file_path = p.resolved[len("file://") :]
 
-                if (
-                    p.type == "CreateFile"
-                ):  # TODO: CreateFile for buckets is not under testing
-                    local_file_path = os.path.join(
-                        file_store.getLocalTempDir(), baseName
-                    )
-                    with open(local_file_path, "wb") as n:
-                        n.write(p.resolved.encode("utf-8"))
+    # TODO: I think we need to recursively populate listings here!
+
+    # Now we need to save all the output files and directories.
+    # We could use a ToilPathMapper, but that contains stuff to work with Toil directories as encoded in Locations, while we can assume that we have all the Files and
 
-                destUrl = "/".join(s.strip("/") for s in [destBucket, baseName])
-                file_store.exportFile(FileID.unpack(local_file_path), destUrl)
+    pm = ToilPathMapper(
+        jobfiles,
+        "",
+        outdir,
+        separateDirs=False,
+        stage_listing=True,
+    )
+    for _, p in pm.items():
+        if p.staged:
+            # We're supposed to copy/expose something.
+            # Note that we have to handle writable versions of everything
+            # because sometimes we might make them in the PathMapper even if
+            # not instructed to copy.
+            if destBucket:
+                # We are saving to a bucket, directories are fake.
+                if p.type in ["File", "CreateFile", "WritableFile", "CreateWritableFile"]:
+                    # Directories don't need to be created if we're exporting to a bucket
+                    baseName = p.target[len(outdir):]
+                    file_id_or_contents = p.resolved
+
+                    if (
+                        p.type in ["CreateFile", "CreateWritableFile"]
+                    ):  # TODO: CreateFile for buckets is not under testing
+
+                        with tempfile.NamedTemporaryFile() as f:
+                            # Make a file with the right contents
+                            f.write(file_id_or_contents.encode('utf-8'))
+                            f.close()
+                            # Import it and pack up the file ID so we can turn around and export it.
+                            file_id_or_contents = 'toilfs:' + toil.importFile(f.name).pack()
+
+                    if file_id_or_contents.startswith('toilfs:'):
+                        # This is something we can export
+                        destUrl = "/".join(s.strip("/") for s in [destBucket, baseName])
+                        toil.exportFile(FileID.unpack(file_id_or_contents[len('toilfs:'):]), destUrl)
+                    # TODO: can a toildir: "file" get here?
```

Comment:
I think if the type is a variant of `File` but the id is `toildir:` then it should error here.  That _should_ never happen.

Reply:
I think it can happen when you take a `toildir:` directory and compose a path to a file within it by joining subdirectories/filenames onto its URL with `/`. That happens in some places, and I'm not sure if it can happen here or not. It might be safest to actually implement `toildir:`.